File: apps/web/src/api/annotationHistory.ts
import { apiClient } from "./client";
import type { UserBrief } from "@/types";

export interface HistoryEntry {
  kind: "audit" | "comment";
  timestamp: string;
  actor: UserBrief | null;
  // audit
  action: string | null;
  detail: Record<string, unknown> | null;
  // comment
  comment_id: string | null;
  body: string | null;
}

export interface AnnotationHistoryResponse {
  // I4 · task 级时间线时 annotation_id=null.
  annotation_id: string | null;
  task_id: string;
  entries: HistoryEntry[];
}

export const annotationHistoryApi = {
  get: (annotationId: string) =>
    apiClient.get<AnnotationHistoryResponse>(
      `/annotations/${annotationId}/history`,
    ),
  // I4 · DiscussionPanel 未选中标注时降级到 task 级时间线.
  getByTask: (taskId: string) =>
    apiClient.get<AnnotationHistoryResponse>(
      `/tasks/${taskId}/audit-history`,
    ),
};

File: apps/api/app/schemas/annotation_history.py
"""v0.7.2 · 标注框编辑/审核历史时间线 schema。"""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel

from app.schemas.user import UserBrief


class HistoryEntry(BaseModel):
    """一条历史事件 — audit 或 comment 统一外形。

    kind ∈ { audit, comment }；audit 含 action + detail；comment 含 body + comment_id。
    所有事件按 timestamp 升序合并。
    """
    kind: str
    timestamp: datetime
    actor: UserBrief | None = None
    # audit 字段
    action: str | None = None
    detail: dict[str, Any] | None = None
    # comment 字段
    comment_id: UUID | None = None
    body: str | None = None


class AnnotationHistoryResponse(BaseModel):
    annotation_id: UUID
    task_id: UUID
    entries: list[HistoryEntry]
